platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,369 @@
+The key of a metric won't change and still be defined as the combination of the `category` and `name`.
+The unique identifier of the metric is the combination key and the labels.
+That is a breaking change since today the key is the unique identifier of a metric.
+Since we do not have an api today to add labels the migration should be minimalistic.
+
+#### Migration of the prometheus endpoint
+
+The `nodeId` of a platform is already added as a label today when using the prometheus endpoint.
+For that endpoint we just need to remove the hard coded `nodeId` label and add a general algorithm to add labels.
+
+#### Migration of the CSV exporter
+
+The CSV exporter use the `nodeId` today to create the file name.
+Since the CSV exporter is only used by the platform module we can continue with that approach.
+
+Let's assume we would have the following metrics:
+
+| Type           | Name   | Category   | labels | Value |
+ |---------------|--------|------------|-------|-------|
+ | Counter          | myCounter1 | private_category | nodeId=1 | 1 |
+ | Counter          | myCounter2 | private_category | nodeId=1 | 1 |
+ | Counter          | myCounter3 | private_category | nodeId=1 | 1 |
+ | Counter          | myCounter4 | private_category | nodeId=1, foo=bar | 1 |
+ | DoubleGauge | myDoubleGauge | public_category | nodeId=1 | 1 |
+ | DoubleGauge | myDoubleGauge | private_category | nodeId=1, foo=bar | 1 |
+        
+
+Today a CSV file is created like this:
+```
+filename:,/var/folders/d8/16q746f12zq6tb1ngwngp8gh0000gn/T/junit15732516610471683378/MainNetStats42.csv,
+myCounter1:,myCounter1,
+myCounter2:,myCounter2,
+myCounter3:,myCounter3,
+myCounter4:,myCounter4,
+myDoubleGauge:,myDoubleGauge,
+myDoubleGauge:,myDoubleGauge,
+
+,,private_category,private_category,private_category,private_category,public_category,private_category,
+,,myCounter1,myCounter2,myCounter3,myCounter4,myDoubleGauge,myDoubleGauge,
+,,1,2,3,4,1.230,4.560,
+```
+
+The first section of the list is metadata that defines the file name (that includes the `nodeId`) and a list of all metric names.
+The metric category is ignored here. As you can see the 2 gauges with same name but different category are ends in 2 times the same line.
+This proposal would not change that behavior.
+
+The second section of the list defines the metrics and their values. This part will be changed to include the labels.
+
+```
+,,private_category,private_category,private_category,private_category,public_category,
+,,myCounter1,myCounter2,myCounter3,myCounter4,myDoubleGauge,
+,,1,2,3,4,1.230,4.560,
+```
+
+Here the first row defines the category of the metric and the second row the name. The 3rd row contains the values.
+An additional row between the categories and values will be added to include the labels of the metric.
+Since a metric can contain multiple labels the cell content must be quoted:
```

Comment:
According to the example, the label row will have a specific format that includes the value.

- what would happen with the total value of the metric (The one aggregating all values without lables)? 
- Is there any tool that parses this file automatically that depends on the format? Is this change still compatible? Who would modify that if not? 

I don't think the file is a well-structured CSV. However, after this modification, parsing the file would become more difficult, as quoting content with a comma character is insufficient to differentiate whether the comma is part of a value or an actual separator.


I wish we could move away from csv files altogether, but from what I know, we need this file for tests,

Have you considered the approach of denormalizing the metric labels? Like:
```
,,private_category,private_category,private_category, private_category
,,metric_with_label,metric_with_label,metric_with_label,metric_no_label
,,"",label_a,label_b,label_c,"", 
10,5,3,2,28
```

the quotes ("") only indicate that the first column of the label represents the total and would be the content for non-labeled metrics.
It seems more consistent with the current format.

Reply:
> what would happen with the total value of the metric (The one aggregating all values without lables)? 

That is never stored. You can only get that by using a prometheus query

> Is there any tool that parses this file automatically that depends on the format?

As far as I know it is only for manual usage. Adding @netopyr and @lpetrovic05 for clarification

> I wish we could move away from csv files altogether, but from what I know, we need this file for tests

I do not think so. As far as I know it is only used by some developers.

> I don't think the file is a well-structured CSV.

I agree on that. Based on that I want to change only what is really needed and hope that we do not use/need it in future anymore. I assume my current idea is "good enough" and whoever uses the file today will ignore the new line anyway

Reply:
At least in the past, the CSV files were used everywhere, and the parsers were not very tolerant. The slightest change broke processes downstream.

Reply:
lol, I always assumed that is the basic out for devs when you do not have grafana access. Let's see if @lpetrovic05 has more knowledge

Reply:
CSV metrics are used to produce PDF graphs when running JRS. It would be great if JRS metrics went to prometheus.

I think the CSVs are also used by some JRS validators, @JeffreyDallas can comment on that

Reply:
ok, so no open source code use them. Based on that I should not add anything to the proposal

Reply:
I don't know why the impact analysis should be limited only to "open source" code. 

It seems that we know:
1)  We have an impact on the CVS file,
2) The parser is used in the execution of JRS tests
3) The parser might not have a lot of tolerance for changes. 

JRS test panels did run on PRs, and I think the nightly test runs with that and uses the metrics' results to compare things. So, they seem pretty important.

If we operate with the premise that the parser can be broken with this change: 
* whose team is responsible for fixing it? 
   If it is platform responsibility, then I think it should be part of this proposal and work at the same time so people can still use the test.
   If it is not the platform's team's responsibility, we should note the impact on the responsible team and get their review, too, as they will have to fix it.

We can also move forward and see what happens along the way.

Reply:
After speaking with Hendrik, he explained why leaving this out of the proposal makes sense. I am resolving this comment.